File: code/ch07/javascript/test_server_stub.js
// §7.5.2 Deploying Test Servers — a stub HTTP server programmed to return 500,
// asserting the clinic client's fallback behavior.
const assert = require("node:assert");
const http = require("node:http");

async function fetchAppointments(baseUrl) {
  const resp = await fetch(baseUrl + "/appointments");
  if (!resp.ok) return [];               // fallback: empty schedule, not a crash
  return Buffer.from(await resp.arrayBuffer());
}

const stub = http.createServer((req, res) => res.writeHead(500).end());
stub.listen(0, "127.0.0.1", async () => {
  const base = `http://127.0.0.1:${stub.address().port}`;
  assert.deepStrictEqual(await fetchAppointments(base), []);
  stub.close();
  console.log("client fell back to an empty schedule on 500");
});

File: code/ch07/ruby/test_server_stub.rb
# §7.5.2 Deploying Test Servers — a stub HTTP server programmed to return 500,
# asserting the clinic client's fallback behavior.
require "net/http"
require "socket"

stub = TCPServer.new("127.0.0.1", 0)
Thread.new do
  loop do
    client = stub.accept                 # a stub programmed to always fail
    client.gets("\r\n\r\n")              # consume the request head
    client.write "HTTP/1.1 500 Internal Server Error\r\nContent-Length: 0\r\n\r\n"
    client.close
  end
end

def fetch_appointments(base_url)
  resp = Net::HTTP.get_response(URI(base_url + "/appointments"))
  resp.is_a?(Net::HTTPSuccess) ? resp.body : []  # fallback: empty schedule
end

port = stub.addr[1]
raise unless fetch_appointments("http://127.0.0.1:#{port}") == []
stub.close
puts "client fell back to an empty schedule on 500"
